### PR TITLE
Added datacontentencoding

### PR DESCRIPTION
**Pull request by clemensv**
Implements content encoding as raised in issue #371

### CODE REVIEW

**Comment by duglin** [comment]
ping @alanconway @n3wscott
Any thoughts on this?

**Comment by alanconway** [comment]
The datacontentencoding changes are all good by me.

On a side note: I'm surprised that critical attribute names are still changing, e.g. "contenttype -> datacontenttype". Also since the spec has adopted the cantdecideonwordseparatorssoletshavenone convention, it seems like a not-so-great idea to be adding more words to multi-word names. However, that's another story.

**Comment by duglin** [comment]
Would an example help clarify the need for this to a newbie?  Could be done in a follow-on PR if so.

File: spec.md [hunk 1]
```diff
@@ -276,6 +274,23 @@ help intermediate gateways determine how to route the events.
     [RFC 2046](https://tools.ietf.org/html/rfc2046)
 * For Media Type examples see [IANA Media Types](http://www.iana.org/assignments/media-types/media-types.xhtml)
 
+### datacontentencoding
+* Type: `String` per [RFC 2045 Section 6.1](https://tools.ietf.org/html/rfc2045#section-6.1)
+* Description: Describes the content encoding for the `data` 
+  attribute for when the `data` field MUST be encoded as a string,
+  like with structured transport binding modes using the JSON event 
+  format, but the `datacontenttype` indicates a non-string media 
+  type. When the `data` field's effective data type is not `String`,
```
Comment (n3wscott) [inline spec.md]: > When the `data` field's effective data type is not `String`,
>  this attribute MUST NOT be set and MUST be ignored when set.

why not? I want to zip my http binary data payload...

Comment (n3wscott) [reply spec.md]: Or I want to base64 encode my binary http payload because I am a weirdo...

Comment (n3wscott) [reply spec.md]: Perhaps not, looking at RFC 2045 I do not see anything I would want to use besides base64.  And normal http transports can give me gzip... 

/lgtm

**Comment by alanconway** [comment]
On Wed, Feb 20, 2019 at 8:38 PM Scott Nichols <notifications@github.com>
wrote:

> *@n3wscott* commented on this pull request.
> ------------------------------
>
> In spec.md
> <https://github.com/cloudevents/spec/pull/387#discussion_r258749647>:
>
> > @@ -276,6 +274,23 @@ help intermediate gateways determine how to route the events.
>      [RFC 2046](https://tools.ietf.org/html/rfc2046)
>  * For Media Type examples see [IANA Media Types](http://www.iana.org/assignments/media-types/media-types.xhtml)
>
> +### datacontentencoding
> +* Type: `String` per [RFC 2045 Section 6.1](https://tools.ietf.org/html/rfc2045#section-6.1)
> +* Description: Describes the content encoding for the `data`
> +  attribute for when the `data` field MUST be encoded as a string,
> +  like with structured transport binding modes using the JSON event
> +  format, but the `datacontenttype` indicates a non-string media
> +  type. When the `data` field's effective data type is not `String`,
>
> Perhaps not, looking at RFC 2045 I do not see anything I would want to use
> besides base64. And normal http transports can give me gzip...
>

Compression is a separate "transfer-encoding" header in HTTP.
Content-encoding is just for transorts that can't store simple binary data,
like 1960s 7-bit-ascii email, or (sigh) JSON in 2019.

File: spec.md [hunk 2]
```diff
@@ -276,6 +274,23 @@ help intermediate gateways determine how to route the events.
...
+  this attribute MUST NOT be set and MUST be ignored when set.
+  
+  The "Base64" value for the Base64 encoding as defined in [RFC 2045 Section 6.8](https://tools.ietf.org/html/rfc2045#section-6.8)
+  MUST be supported. When set, the event-format-encoded value of the `data` 
+  attribute is a base64 string, but the effective data type of 
+  the `data` attribute towards the application is the base64-decoded
+  binary array.
+  
+  All other RFC2045 schemes are undefined for CloudEvents.
```
Comment (evankanderson) [inline spec.md]: Add a "Constraints" section, like the other attributes?

I think at the moment, "Constraints" is "Must be the string 'Base64'". If so, would it make more sense to standardize on assuming 'Base64' encoding and excluding the other options?

Comment (duglin) [reply spec.md]: @clemensv  ^^^

**Comment by alanconway** [comment]
perhaps "contentencoding" rather than "datacontentencoding"? Given the
(unfortunate IMO) choice of noseparatorswhatsoever for attribute names, I
think it's important to try hard for names that are only one word, and
absolutely ban names that are more than two words.

**Comment by clemensv** [comment]
@alanconway we chose to rename contenttype to datacontenttype because people were confused about what that relates to, and now this also needs to have data in front of it, because it relates to the same attribute.

**Comment by duglin** [comment]
All - This one has been updated - please review and comment

**Comment by rperelma** [comment]
Might be good to add an example

**Comment by JemDay** [comment]
+1 for some examples - but maybe there should be a separate PR to add a few examples based on recent changes.

Generally LGTM but deferring final comment until merge conflicts are resolved.

**Comment by duglin** [comment]
Approved on the 4/4 call

**Pull request merged**